userd/autostart.go

```diff
@@ -0,0 +1,210 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+// expandDesktopFields processes the input string and expands any %<char>
+// patterns. '%%' expands to '%', all other patterns expand to empty strings.
+func expandDesktopFields(in string) string {
+	raw := []rune(in)
+	out := make([]rune, 0, len(raw))
+
+	var hasKey bool
+	for _, r := range raw {
+		if hasKey {
+			hasKey = false
+			// only allow %% -> % expansion, drop other keys
+			if r == '%' {
+				out = append(out, r)
+			}
+			continue
+		} else if r == '%' {
+			hasKey = true
+			continue
+		}
+		out = append(out, r)
+	}
+	return string(out)
+}
+
+func findExec(desktopFileContent []byte) (string, error) {
+	scanner := bufio.NewScanner(bytes.NewBuffer(desktopFileContent))
+	execCmd := ""
+	for scanner.Scan() {
+		bline := scanner.Bytes()
+
+		if !bytes.HasPrefix(bline, []byte("Exec=")) {
+			continue
+		}
+
+		full := string(bline[len("Exec="):])
+		execCmd = expandDesktopFields(full)
+		break
+	}
+
+	execCmd = strings.TrimSpace(execCmd)
+	if execCmd == "" {
+		return "", fmt.Errorf("Exec not found or invalid")
+	}
+	return execCmd, nil
+}
+
+func getCurrentSnapInfo(snapName string) (*snap.Info, error) {
```

Comment:
Should this live in `snap/ somewhere?

Reply:
moved as `snap.ReadCurrentInfo()`, also updated a callsite in `cmd/snap/cmd_run.go`